File: backend-rust/src/grpc/task.rs
use std::sync::Arc;
use std::time::Instant;

use chrono::{NaiveDate, NaiveDateTime, Utc};
use tonic::{Request, Response, Status};

use crate::auth::{resolve_user_id, ResolveOutcome};
use crate::db;
use crate::model::{validate_task_input, Task, TaskInput, TaskStatus};
use crate::AppState;

use super::pb;
use pb::task_service_server::TaskService;

pub struct TaskGrpcService {
    pub state: Arc<AppState>,
}

fn naive_to_timestamp(dt: NaiveDateTime) -> prost_types::Timestamp {
    let utc = dt.and_utc();
    prost_types::Timestamp { seconds: utc.timestamp(), nanos: utc.timestamp_subsec_nanos() as i32 }
}

fn task_to_pb(task: &Task) -> pb::Task {
    pb::Task {
        id: task.id,
        name: task.name.clone(),
        description: task.description.clone(),
        status: task.status.as_str().to_string(),
        finished_on: task.finished_on.format("%Y-%m-%d").to_string(),
        labels: task
            .labels
            .iter()
            .map(|l| pb::Label { id: l.id, name: l.name.clone() })
            .collect(),
        created_at: Some(naive_to_timestamp(task.created_at)),
        updated_at: Some(naive_to_timestamp(task.updated_at)),
    }
}

/// backend(Go)のTaskServer.resolveUserID(gRPC v2版)と同じ:
/// メタデータのauthorizationを検証し、未検証/未プロビジョニングをgRPCステータスへ変換する
async fn authenticate<T>(state: &AppState, req: &Request<T>) -> Result<u64, Status> {
    let auth_header = req
        .metadata()
        .get("authorization")
        .and_then(|v| v.to_str().ok())
        .ok_or_else(|| Status::unauthenticated("unauthorized"))?;
    let token = auth_header
        .strip_prefix("Bearer ")
        .ok_or_else(|| Status::unauthenticated("unauthorized"))?;

    let claims = state
        .dispatcher
        .verify(token)
        .await
        .map_err(|_| Status::unauthenticated("unauthorized"))?;

    match resolve_user_id(&state.pool, &claims).await {
        ResolveOutcome::Ok(user_id) => Ok(user_id),
        ResolveOutcome::UserNotProvisioned => Err(Status::permission_denied("user not provisioned")),
    }
}

fn parse_finished_on(s: &str) -> Result<NaiveDate, Status> {
    NaiveDate::parse_from_str(s, "%Y-%m-%d").map_err(|e| Status::invalid_argument(format!("invalid finished_on: {}", e)))
}

fn build_input(name: String, description: Option<String>, status: String, finished_on: &str, label_ids: Vec<u64>) -> Result<(TaskInput, NaiveDate), Status> {
    let parsed = parse_finished_on(finished_on)?;
    Ok((TaskInput { name, description, status, finished_on: parsed, label_ids }, parsed))
}

fn validate(input: &TaskInput) -> Result<TaskStatus, Status> {
    let today = Utc::now().date_naive();
    validate_task_input(input, today).map_err(Status::invalid_argument)
}

#[tonic::async_trait]
impl TaskService for TaskGrpcService {
    async fn list_tasks(
        &self,
        request: Request<pb::ListTasksRequest>,
    ) -> Result<Response<pb::ListTasksResponse>, Status> {
        let user_id = authenticate(&self.state, &request).await?;
        let req = request.into_inner();

        let status = if req.status.is_empty() {
            None
        } else {
            Some(TaskStatus::from_str(&req.status).ok_or_else(|| Status::invalid_argument(format!("invalid status: {}", req.status)))?)
        };
        let limit = if req.limit <= 0 { 20 } else { req.limit as i64 };

        let tasks = db::list_tasks_cursor(&self.state.pool, user_id, &req.name, status, &req.label_ids, req.cursor, limit)
            .await
            .map_err(|e| Status::internal(format!("list tasks: {}", e)))?;

        let next_cursor = if (tasks.len() as i64) < limit {
            0
        } else {
            tasks.last().map(|t| t.id).unwrap_or(0)
        };

        Ok(Response::new(pb::ListTasksResponse {
            tasks: tasks.iter().map(task_to_pb).collect(),
            next_cursor,
        }))
    }

    async fn get_task(&self, request: Request<pb::GetTaskRequest>) -> Result<Response<pb::Task>, Status> {
        let user_id = authenticate(&self.state, &request).await?;
        let id = request.into_inner().id;
        let task = db::get_task(&self.state.pool, id, user_id)
            .await
            .map_err(|e| Status::internal(e.to_string()))?
            .ok_or_else(|| Status::not_found("task not found"))?;
        Ok(Response::new(task_to_pb(&task)))
    }

    async fn create_task(&self, request: Request<pb::CreateTaskRequest>) -> Result<Response<pb::Task>, Status> {
        let user_id = authenticate(&self.state, &request).await?;
        let req = request.into_inner();
        let (input, _) = build_input(req.name, req.description, req.status, &req.finished_on, req.label_ids)?;
        let status = validate(&input)?;

        let id = db::create_task(&self.state.pool, user_id, &input, status)
            .await
            .map_err(|e| Status::internal(e.to_string()))?;
        let task = db::get_task(&self.state.pool, id, user_id)
            .await
            .map_err(|e| Status::internal(e.to_string()))?
            .ok_or_else(|| Status::internal("task disappeared after create"))?;
        Ok(Response::new(task_to_pb(&task)))
    }

    async fn update_task(&self, request: Request<pb::UpdateTaskRequest>) -> Result<Response<pb::Task>, Status> {
        let user_id = authenticate(&self.state, &request).await?;
        let req = request.into_inner();
        let (input, _) = build_input(req.name, req.description, req.status, &req.finished_on, req.label_ids)?;
        let status = validate(&input)?;

        let updated = db::update_task(&self.state.pool, req.id, user_id, &input, status)
            .await
            .map_err(|e| Status::internal(e.to_string()))?;
        if !updated {
            return Err(Status::not_found("task not found"));
        }
        let task = db::get_task(&self.state.pool, req.id, user_id)
            .await
            .map_err(|e| Status::internal(e.to_string()))?
            .ok_or_else(|| Status::not_found("task not found"))?;
        Ok(Response::new(task_to_pb(&task)))
    }

    async fn delete_task(&self, request: Request<pb::DeleteTaskRequest>) -> Result<Response<pb::DeleteTaskResponse>, Status> {
        let user_id = authenticate(&self.state, &request).await?;
        let id = request.into_inner().id;
        let deleted = db::delete_task(&self.state.pool, id, user_id)
            .await
            .map_err(|e| Status::internal(e.to_string()))?;
        if !deleted {
            return Err(Status::not_found("task not found"));
        }
        Ok(Response::new(pb::DeleteTaskResponse {}))
    }
}

/// method/実際のgRPCステータス(Status::code())/durationを1rpc1行のINFOログとして出す
///
/// 【実機検証で判明・修正】以前は`GrpcLoggingLayer`(tower::Layer、HTTP/tonicのトランスポート層を
/// ラップする実装)でmethod+durationだけログしていたが、実際のgRPCステータス(成否)はレスポンスの
/// HTTP/2トレーラーに乗るため、トランスポート層からは素直に取れなかった(トレーラー到達を検知する
/// bodyラップが別途必要になる)。
///
/// ここではその代わりに、TaskServiceトレイトの各メソッドが返す`Result<Response<T>, Status>`を
/// そのまま見る(Goの`grpc.ChainUnaryInterceptor`がhandlerの`err`をそのまま受け取れるのと同じ発想、
/// Railsの`GRPC::BadStatus#code`と同じ発想)。この層は既にビジネスロジックの戻り値そのものであり、
/// トレーラーを待つ必要が無いため、実際のステータスコードを正確に記録できる
async fn logged<T>(method: &str, fut: impl std::future::Future<Output = Result<Response<T>, Status>>) -> Result<Response<T>, Status> {
    let start = Instant::now();
    let result = fut.await;
    let duration_ms = start.elapsed().as_millis();
    let code = match &result {
        Ok(_) => tonic::Code::Ok,
        Err(status) => status.code(),
    };
    tracing::info!(method, status = ?code, duration_ms, "grpc request");
    result
}

/// TaskGrpcServiceを包み、全rpcにlogged()を通す薄いデコレータ
/// (Rustにはモンキーパッチが無いため、5メソッド分のログをtrait実装レベルで一箇所にまとめる手段として
/// デコレータ構造体を使う。Railsの`Module#prepend`・Goのinterceptorチェーンと同じ役割)
pub struct LoggingTaskGrpcService {
    pub inner: TaskGrpcService,
}

#[tonic::async_trait]
impl TaskService for LoggingTaskGrpcService {
    async fn list_tasks(
        &self,
        request: Request<pb::ListTasksRequest>,
    ) -> Result<Response<pb::ListTasksResponse>, Status> {
        logged("list_tasks", self.inner.list_tasks(request)).await
    }

    async fn get_task(&self, request: Request<pb::GetTaskRequest>) -> Result<Response<pb::Task>, Status> {
        logged("get_task", self.inner.get_task(request)).await
    }

    async fn create_task(&self, request: Request<pb::CreateTaskRequest>) -> Result<Response<pb::Task>, Status> {
        logged("create_task", self.inner.create_task(request)).await
    }

    async fn update_task(&self, request: Request<pb::UpdateTaskRequest>) -> Result<Response<pb::Task>, Status> {
        logged("update_task", self.inner.update_task(request)).await
    }

    async fn delete_task(&self, request: Request<pb::DeleteTaskRequest>) -> Result<Response<pb::DeleteTaskResponse>, Status> {
        logged("delete_task", self.inner.delete_task(request)).await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // 【テスト監査で追加】LoggingTaskGrpcServiceの5メソッドは全てlogged()を経由するだけの
    // 薄いラップなので、logged()自体が「戻り値・エラーをそのまま透過させる」ことさえ
    // 保証できれば、実質的にデコレータ全体の正しさを裏付けられる(5メソッド分の重複テスト
    // を書かずに済む)。以前は grpc/task.rs にテストが1件も無く、
    // LoggingTaskGrpcService(main.rsで実際に使われる方)を経由する経路が
    // 一度も検証されていなかった。
    #[tokio::test]
    async fn logged_passes_through_ok_response_unchanged() {
        let result = logged("list_tasks", async { Ok(Response::new(42u32)) }).await;
        let inner = result.expect("Okがloggingを通してErrに化けている").into_inner();
        assert_eq!(inner, 42, "レスポンスの中身がラップ前後で変わっている");
    }

    #[tokio::test]
    async fn logged_passes_through_error_status_unchanged() {
        // 実際のハンドラ(get_task等)がNotFoundを返すケースを模す
        let result: Result<Response<()>, Status> =
            logged("get_task", async { Err(Status::not_found("task not found")) }).await;
        let status = result.expect_err("Errがloggingを通してOkに化けている(呼び出し元がエラーを検知できなくなる実害)");
        assert_eq!(status.code(), tonic::Code::NotFound, "ステータスコードが変わっている");
        assert_eq!(status.message(), "task not found", "エラーメッセージが変わっている");
    }

    #[tokio::test]
    async fn logged_does_not_panic_for_any_status_code() {
        // status = ?code (Debugフォーマット)が、tonic::Codeの全バリアントで
        // panicせず記録できることを確認する(将来tonicがバリアントを追加した場合の
        // 回帰検知も兼ねる)
        let codes = [
            tonic::Code::Cancelled,
            tonic::Code::Unknown,
            tonic::Code::InvalidArgument,
            tonic::Code::DeadlineExceeded,
            tonic::Code::NotFound,
            tonic::Code::AlreadyExists,
            tonic::Code::PermissionDenied,
            tonic::Code::ResourceExhausted,
            tonic::Code::FailedPrecondition,
            tonic::Code::Aborted,
            tonic::Code::OutOfRange,
            tonic::Code::Unimplemented,
            tonic::Code::Internal,
            tonic::Code::Unavailable,
            tonic::Code::DataLoss,
            tonic::Code::Unauthenticated,
        ];
        for code in codes {
            let result: Result<Response<()>, Status> =
                logged("x", async { Err(Status::new(code, "msg")) }).await;
            assert_eq!(result.unwrap_err().code(), code);
        }
    }
}

File: backend-js-express/src/grpc/task.js
'use strict';

const grpc = require('@grpc/grpc-js');
const db = require('../db');
const { validateTaskInput, todayUtcIso, statusToString, statusFromString } = require('../model');
const { resolveUserId } = require('../auth');
const { logInfo } = require('../logging');

function grpcError(code, message) {
  const err = new Error(message);
  err.code = code;
  err.details = message;
  return err;
}

// "YYYY-MM-DD HH:MM:SS"(UTC、dateStrings:trueで取得した文字列) -> google.protobuf.Timestamp
function toTimestamp(mysqlDatetime) {
  const ms = Date.parse(`${mysqlDatetime.replace(' ', 'T')}Z`);
  return { seconds: Math.floor(ms / 1000), nanos: 0 };
}

function taskToPb(task) {
  return {
    id: task.id,
    name: task.name,
    description: task.description ?? undefined,
    status: statusToString(task.status),
    finished_on: task.finishedOn,
    labels: task.labels.map((l) => ({ id: l.id, name: l.name })),
    created_at: toTimestamp(task.createdAt),
    updated_at: toTimestamp(task.updatedAt),
  };
}

// backend(Go)のTaskServer.resolveUserID(gRPC v2版)・backend-rustのauthenticateと同じ:
// メタデータのauthorizationを検証し、未検証/未プロビジョニングをgRPCステータスへ変換する
async function authenticate(state, call) {
  const metadata = call.metadata.getMap();
  const authHeader = metadata.authorization;
  if (!authHeader) throw grpcError(grpc.status.UNAUTHENTICATED, 'unauthorized');
  const m = /^Bearer (.+)$/.exec(authHeader);
  if (!m) throw grpcError(grpc.status.UNAUTHENTICATED, 'unauthorized');

  let claims;
  try {
    claims = await state.dispatcher.verify(m[1]);
  } catch (e) {
    throw grpcError(grpc.status.UNAUTHENTICATED, 'unauthorized');
  }

  const resolved = await resolveUserId(state.pool, claims);
  if (!resolved.ok) throw grpcError(grpc.status.PERMISSION_DENIED, 'user not provisioned');
  return resolved.userId;
}

function parseFinishedOn(s) {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(s || '')) {
    throw grpcError(grpc.status.INVALID_ARGUMENT, `invalid finished_on: ${s}`);
  }
  return s;
}

function buildInput(req) {
  const finishedOn = parseFinishedOn(req.finished_on);
  return {
    name: req.name || '',
    description: req.description || null,
    status: req.status || '',
    finishedOn,
    labelIds: (req.label_ids || []).map(Number),
  };
}

function validate(input) {
  const result = validateTaskInput(input, todayUtcIso());
  if (result.error) throw grpcError(grpc.status.INVALID_ARGUMENT, result.error);
  return result.status;
}

function toGrpcError(e) {
  if (e && typeof e.code === 'number') return e;
  return grpcError(grpc.status.INTERNAL, e && e.message ? e.message : 'internal error');
}

function buildService(state) {
  return {
    async listTasks(call, callback) {
      try {
        const userId = await authenticate(state, call);
        const req = call.request;
        let status = null;
        if (req.status) {
          status = statusFromString(req.status);
          if (status === null) throw grpcError(grpc.status.INVALID_ARGUMENT, `invalid status: ${req.status}`);
        }
        const limit = !req.limit || req.limit <= 0 ? 20 : req.limit;
        const tasks = await db.listTasksCursor(
          state.pool,
          userId,
          req.name || '',
          status,
          (req.label_ids || []).map(Number),
          req.cursor || 0,
          limit,
        );
        const nextCursor = tasks.length < limit ? 0 : tasks[tasks.length - 1].id;
        callback(null, { tasks: tasks.map(taskToPb), next_cursor: nextCursor });
      } catch (e) {
        callback(toGrpcError(e));
      }
    },

    async getTask(call, callback) {
      try {
        const userId = await authenticate(state, call);
        const task = await db.getTask(state.pool, Number(call.request.id), userId);
        if (!task) throw grpcError(grpc.status.NOT_FOUND, 'task not found');
        callback(null, taskToPb(task));
      } catch (e) {
        callback(toGrpcError(e));
      }
    },

    async createTask(call, callback) {
      try {
        const userId = await authenticate(state, call);
        const input = buildInput(call.request);
        const status = validate(input);
        const id = await db.createTask(state.pool, userId, input, status);
        const task = await db.getTask(state.pool, id, userId);
        callback(null, taskToPb(task));
      } catch (e) {
        callback(toGrpcError(e));
      }
    },

    async updateTask(call, callback) {
      try {
        const userId = await authenticate(state, call);
        const id = Number(call.request.id);
        const input = buildInput(call.request);
        const status = validate(input);
        const updated = await db.updateTask(state.pool, id, userId, input, status);
        if (!updated) throw grpcError(grpc.status.NOT_FOUND, 'task not found');
        const task = await db.getTask(state.pool, id, userId);
        callback(null, taskToPb(task));
      } catch (e) {
        callback(toGrpcError(e));
      }
    },

    async deleteTask(call, callback) {
      try {
        const userId = await authenticate(state, call);
        const id = Number(call.request.id);
        const deleted = await db.deleteTask(state.pool, id, userId);
        if (!deleted) throw grpcError(grpc.status.NOT_FOUND, 'task not found');
        callback(null, {});
      } catch (e) {
        callback(toGrpcError(e));
      }
    },
  };
}

function grpcCodeName(code) {
  const entry = Object.entries(grpc.status).find(([, v]) => v === code);
  return entry ? entry[0] : String(code);
}

// method/実際のgRPCステータス/durationを1rpc1行のログとして出す(backend-rustのgrpc/task.rsの
// LoggingTaskGrpcServiceと同じ方針: ハンドラが返すエラー/成功をそのまま見るので、
// HTTP/2トレーラーを覗く必要が無く正確なステータスを記録できる)
function withLogging(service) {
  const wrapped = {};
  for (const [method, fn] of Object.entries(service)) {
    wrapped[method] = (call, callback) => {
      const start = process.hrtime.bigint();
      fn(call, (err, response) => {
        const durationMs = Number(process.hrtime.bigint() - start) / 1e6;
        const code = err ? (typeof err.code === 'number' ? err.code : grpc.status.UNKNOWN) : grpc.status.OK;
        logInfo('grpc request', { method, status: grpcCodeName(code), duration_ms: Math.round(durationMs) });
        callback(err, response);
      });
    };
  }
  return wrapped;
}

module.exports = { buildService, withLogging, taskToPb, toTimestamp };
